### DISCUSSION TITLE
Conditions that may cause UnroutablePeer events (since libp2p 0.52 update)

### DISCUSSION THREAD

**Discussion by b-zee** [libp2p/rust-libp2p#4290] [General]
In a recent test we did with people running nodes from home, there were a few occurrences of Kademlia nodes that were seen by others as [`UnroutablePeer`](https://docs.rs/libp2p/latest/libp2p/kad/enum.KademliaEvent.html#variant.UnroutablePeer). In our codebase, we haven't done much that I would think causes this, so I am looking at some potential changes in libp2p.

`UnroutablePeer` seems to be emitted when the peer has dialed us, instead of us dialing them ([here](https://github.com/libp2p/rust-libp2p/blob/f95f39f634d08dad8df72d7ecf92312f9ad2d826/protocols/kad/src/behaviour.rs#L2158) the address is `None` and that causes the event later on). The idea behind this is solid, as we do not want to add nodes that are not reachable/routable.

<s>I have a hunch this has to do with my PR #4224, causing us not to dial outwards for new nodes and thus never being able to tell if a peer is reachable/routable. (Idea: a new condition to dial only if we do not have an outgoing connection already.)</s>

But, perhaps there are more conditions that this can occur in. Perhaps there is something going on with AutoNAT or external addresses or with the client/server mode.

**Comment by dariusc93** [comment]
The PR only prevents dialing a peer that is already being dialed (which would prevent redundant dialing). Do you have any code or logs you could provide to be able to determine the actual cause?

**Reply by b-zee** [reply]
> The PR only prevents dialing a peer that is already being dialed

You're right. It doesn't seem related actually.

**Comment by thomaseizinger** [comment]
> (Idea: a new condition to dial only if we do not have an outgoing connection already.)

You should be able to do this yourself by building a `NetworkBehaviour` that tracks your outbound connections and dials nodes as you feed addresses to it. At the moment, we don't yet have a mechanism for `NetworkBehaviour`s to exchange addresses of other nodes. Similar to https://github.com/libp2p/rust-libp2p/pull/3651, we could add a message to `ToSwarm` that allows `NetworkBehaviour`s to share such addresses.

Identify could use that to share the self-reported addresses of a node with other behaviours and your custom behaviour could pick that up and attempt to dial such an address. In case the node has a public address, it would then be added to the routing table.

We also have UPnP in the works which might help for deployments in people's home's to automatically perform a port-forwarding: https://github.com/libp2p/rust-libp2p/pull/4156

**Reply by mxinden** [reply]
> > (Idea: a new condition to dial only if we do not have an outgoing connection already.)
> 
> You should be able to do this yourself by building a `NetworkBehaviour` that tracks your outbound connections and dials nodes as you feed addresses to it. At the moment, we don't yet have a mechanism for `NetworkBehaviour`s to exchange addresses of other nodes. Similar to #3651, we could add a message to `ToSwarm` that allows `NetworkBehaviour`s to share such addresses.

Created a feature request. @thomaseizinger @b-zee what do you think? https://github.com/libp2p/rust-libp2p/issues/4302
